examples/src/Crowdfunding/Offchain.hs

```diff
@@ -0,0 +1,345 @@
+-- | Individual can request a refund
+txRefund :: MonadBlockChain m => m ()
+txRefund = do
+  funder <- ownPaymentPubKeyHash
+  utxos <-
+    scriptUtxosSuchThat Cf.crowdfundingValidator (\d _ -> Cf.getFunder d == Just funder)
+  void $  validateTxSkel $ txSkel $
+    map (SpendsScript Cf.crowdfundingValidator Cf.IndividualRefund) utxos
+      :=>:
+    [ paysPK funder (PlutusTx.Prelude.sum $ map (sOutValue . fst) utxos) ]
+
+-- | Owner can fund the project before the deadline. When funding, mint n reward tokens,
+-- one for each funder. Each token should go to the address of the utxo of the funders.
+txProjectFund :: (MonadBlockChain m) => Cf.ValParams -> m ()
+txProjectFund p = do
+  fundingTarget <- ownPaymentPubKeyHash
+  utxos <-
+    scriptUtxosSuchThat Cf.crowdfundingValidator (\d _ -> Cf.getOwner d == fundingTarget)
+  let datumTotal = PlutusTx.Prelude.sum $ mapMaybe (Cf.getValue . snd) utxos
+      q = Cf.PolicyParams { Cf.pRewardTokenName = Cf.rewardTokenName fundingTarget }
+      uniqueAddrs = nub $ mapMaybe (Cf.getFunder . snd) utxos
+      token num = Value.assetClassValue (Cf.rewardTokenAssetClass p) num
+  void $ validateTxSkel $ txSkel $
+    ( Before (Cf.projectDeadline p) :
+      Mints
+        (Just (Scripts.validatorAddress Cf.crowdfundingValidator))
+        [Cf.rewardTokenPolicy q]
+        (token $ fromIntegral $ length uniqueAddrs) :
+      map (SpendsScript Cf.crowdfundingValidator Cf.Launch) utxos
+    )
+      :=>:
+    ( paysPK fundingTarget (datumTotal <> minAda) :
+      map (`paysPK` (token 1 <> minAda)) uniqueAddrs
+      -- uncomment below (and comment above) to attempt to pay owner all tokens
+      -- map (\_ -> paysPK fundingTarget token) uniqueAddrs
+    )
+
+-- | Get total contributions from a specific address given all utxos
+getContributionsAddr :: [(SpendableOut, Cf.Datum)] -> L.PubKeyHash -> L.Value
+getContributionsAddr utxos addr =
+  let addrUtxos = filter (\utxo -> Cf.getFunder (snd utxo) == Just addr) utxos
+  in PlutusTx.Prelude.sum $ map (sOutValue . fst) addrUtxos
+
+-- | Owner can refund all contributors after the deadline
+txRefundAll :: (MonadBlockChain m) => Cf.ValParams -> m ()
+txRefundAll p = do
+  fundingTarget <- ownPaymentPubKeyHash
+  utxos <-
+    scriptUtxosSuchThat Cf.crowdfundingValidator (\d _ -> Cf.getOwner d == fundingTarget)
+  -- First, gather all contributions from the same address. This confirms that we only use
+  -- one output to pay back a contributor, even if they contributed multiple times.
+  let addrUtxos = filter (isJust . Cf.getFunder . snd) utxos
+      uniqueAddrs = nub $ mapMaybe (Cf.getFunder . snd) utxos
+      contributions = map (getContributionsAddr addrUtxos) uniqueAddrs
+  void $ validateTxSkel $ txSkel $
+    ( After (Cf.projectDeadline p) :
+      map (SpendsScript Cf.crowdfundingValidator Cf.Launch) utxos
+    )
+      :=>:
+    zipWith paysPK uniqueAddrs contributions
+        
+-- Just so we have something to fund that's not Ada:
+-- Have a banana
+
```

Comment:
I think that the contents of this file from this point onwards should be reafactored into a `tests/CrowdfundingSpec.hs`